File: template-api-note-writer/src/note_writer/enhanced_research.py
import os
from typing import List, Optional, Tuple

from data_models import Post, NoteResult, ProposedMisleadingNote, MisleadingTag
from note_writer.openrouter_util import (
    perplexity_search,
    identify_missing_context,
    evaluate_source_trustworthiness,
    extract_urls_from_text,
    fetch_webpage_content,
    find_context_in_source,
    claude_analyze
)


def _get_search_query_for_post(post: Post, images_summary: str) -> str:
    """
    Generate a search query based on the post content for fact-checking.
    """
    prompt = f"""Based on this X post, generate a search query to fact-check its claims and find relevant context.

Post text:
```
{post.text}
```

Images summary:
```
{images_summary if images_summary else "No images"}
```

Generate a concise search query that will help find:
1. Facts related to the main claims
2. Official sources or primary documents
3. Recent developments or updates

Return only the search query, nothing else.
"""
    
    return claude_analyze(prompt, temperature=0.0)


def _format_note_with_context(missing_context: str, source_url: str) -> str:
    """
    Format the final Community Note with missing context and source.
    Must be under 280 characters.
    """
    # Try to create a concise note
    note = f"Missing context: {missing_context} {source_url}"
    
    # If too long, shorten the context description
    if len(note) > 280:
        # Calculate how much we need to shorten
        excess = len(note) - 280 + 3  # +3 for "..."
        shortened_context = missing_context[:len(missing_context) - excess] + "..."
        note = f"Missing context: {shortened_context} {source_url}"
    
    return note


def enhanced_research_and_write_note(
    post: Post,
    images_summary: str = ""
) -> NoteResult:
    """
    Enhanced research pipeline using Perplexity search and Claude analysis.
    """
    try:
        # Phase 1: Initial Research with Perplexity
        search_query = _get_search_query_for_post(post, images_summary)
        search_results = perplexity_search(
            f"Search for information about: {search_query}. "
            f"Focus on authoritative sources and recent information. "
            f"Include specific URLs with each piece of information."
        )
        
        # Phase 2: Identify Missing Context
        missing_context = identify_missing_context(
            post.text,
            images_summary,
            search_results
        )
        
        # Check if no context is missing
        if "NO MISSING CONTEXT" in missing_context:
            return NoteResult(
                post=post,
                refusal="NO NOTE NEEDED - No important missing context identified",
                images_summary=images_summary
            )
        
        # Phase 3: Extract and evaluate sources
        urls = extract_urls_from_text(search_results)
        if not urls:
            return NoteResult(
                post=post,
                refusal="NOT ENOUGH EVIDENCE TO WRITE A GOOD COMMUNITY NOTE - No sources found",
                images_summary=images_summary
            )
        
        # Create a formatted list of sources for trustworthiness evaluation
        sources_text = "\n".join([f"- {url}" for url in urls])
        trust_evaluation = evaluate_source_trustworthiness(sources_text)
        
        # Extract the most trusted source URL from the evaluation
        # This is a simple extraction - in production you'd parse more carefully
        lines = trust_evaluation.strip().split('\n')
        most_trusted_url = None
        for line in lines:
            if 'http' in line:
                # Extract URL from line
                for url in urls:
                    if url in line:
                        most_trusted_url = url
                        break
                if most_trusted_url:
                    break
        
        if not most_trusted_url:
            # Fallback to first URL if parsing fails
            most_trusted_url = urls[0]
        
        # Phase 4: Fetch content from most trusted source
        source_content = fetch_webpage_content(most_trusted_url)
        if not source_content:
            # Try next best source if available
            if len(urls) > 1:
                source_content = fetch_webpage_content(urls[1])
                most_trusted_url = urls[1]
        
        if not source_content:
            return NoteResult(
                post=post,
                refusal="NOT ENOUGH EVIDENCE TO WRITE A GOOD COMMUNITY NOTE - Could not fetch source content",
                images_summary=images_summary
            )
        
        # Phase 5: Find specific context in source
        context_verification = find_context_in_source(
            source_content,
            most_trusted_url,
            missing_context
        )
        
        if "SOURCE DOES NOT ADDRESS CONTEXT" in context_verification:
            # Try with other sources if available
            for url in urls[1:3]:  # Try up to 2 more sources
                source_content = fetch_webpage_content(url)
                if source_content:
                    context_verification = find_context_in_source(
                        source_content,
                        url,
                        missing_context
                    )
                    if "SOURCE DOES NOT ADDRESS CONTEXT" not in context_verification:
                        most_trusted_url = url
                        break
        
        # If still no good source, return refusal
        if "SOURCE DOES NOT ADDRESS CONTEXT" in context_verification:
            return NoteResult(
                post=post,
                refusal="NOT ENOUGH EVIDENCE TO WRITE A GOOD COMMUNITY NOTE - Sources don't adequately address missing context",
                images_summary=images_summary
            )
        
        # Create the note
        # Extract the first/most important missing context point
        context_lines = missing_context.strip().split('\n')
        primary_context = context_lines[0].strip()
        if primary_context.startswith('1.') or primary_context.startswith('-'):
            primary_context = primary_context[2:].strip()
        
        note_text = _format_note_with_context(primary_context, most_trusted_url)
        
        # Determine appropriate misleading tag
        # Since we're focusing on missing context, this will usually be "missing_important_context"
        misleading_tag = MisleadingTag.missing_important_context
        
        return NoteResult(
            post=post,
            note=ProposedMisleadingNote(
                post_id=post.post_id,
                note_text=note_text,
                misleading_tags=[misleading_tag]
            ),
            images_summary=images_summary
        )
        
    except Exception as e:
        return NoteResult(
            post=post,
            error=f"Error in enhanced research pipeline: {str(e)}",
            images_summary=images_summary
        )


def _simple_prompt_note_writing(post: Post, images_summary: str, search_results: str) -> str:
    """
    Alternative approach: Let Claude write the entire note based on research.
    This is a fallback option that's more similar to the original approach.
    """
    prompt = f"""Based on this X post and search results, write a Community Note if important context is missing.

The note should:
- Focus on missing context that materially changes understanding
- Be under 280 characters
- Include exactly one trusted source URL
- Be written neutrally for broad political acceptance
- Start with "Missing context:" if context is needed

If no important context is missing, respond with "NO NOTE NEEDED"

Post text:
```
{post.text}
```

Images:
```
{images_summary if images_summary else "No images"}
```

Search results:
```
{search_results}
```

Write the note or "NO NOTE NEEDED":
"""
    
    return claude_analyze(prompt)

File: community-notes-ts/src/services/enhanced-research.ts
import axios from 'axios';
import { JSDOM } from 'jsdom';
import { Readability } from '@mozilla/readability';
import { Post, NoteResult, ProposedMisleadingNote, MisleadingTag, Config } from '../types';
import { OpenRouterClient } from '../lib/openrouter-client';
import { SourceTrustCacheService } from './source-trust-cache';

export class EnhancedResearchService {
  private openRouter: OpenRouterClient;
  private config: Config;
  private trustCache: SourceTrustCacheService;

  constructor(config: Config) {
    this.config = config;
    this.openRouter = new OpenRouterClient(config);
    this.trustCache = new SourceTrustCacheService();
  }

  async researchAndWriteNote(post: Post, imagesSummary: string = ''): Promise<NoteResult> {
    try {
      // Phase 1: Perform Perplexity search directly with the tweet text
      const searchPrompt = `Find context and information about this tweet: "${post.text}"${imagesSummary ? `\n\nImages in tweet: ${imagesSummary}` : ''}\n\nFocus on authoritative sources and recent information. Include specific URLs with each piece of information.`;
      const searchResults = await this.openRouter.perplexitySearch(searchPrompt);

      // Phase 2: Identify missing context
      const missingContext = await this.openRouter.identifyMissingContext(
        post.text,
        imagesSummary,
        searchResults
      );

      // Check if no context is missing
      if (missingContext.includes('NO MISSING CONTEXT')) {
        return {
          post,
          refusal: 'NO NOTE NEEDED - No important missing context identified',
          images_summary: imagesSummary
        };
      }

      // Phase 3: Extract URLs from the missing context claims (not the original search)
      const claimUrls = this.extractUrlsFromClaims(missingContext);
      if (claimUrls.length === 0) {
        return {
          post,
          refusal: 'NOT ENOUGH EVIDENCE TO WRITE A GOOD COMMUNITY NOTE - No sources found in claims',
          images_summary: imagesSummary
        };
      }

      // Check cache first
      const cachedScores = this.trustCache.getCachedScores(claimUrls);
      const uncachedUrls = this.trustCache.filterUncachedUrls(claimUrls);
      
      // Evaluate only uncached URLs
      let evaluatedSources: { url: string; score: number; reason: string }[] = [];
      if (uncachedUrls.length > 0) {
        const sourcesText = uncachedUrls.map(url => `- ${url}`).join('\n');
        const trustEvaluation = await this.openRouter.evaluateSourceTrustworthiness(sourcesText);
        
        // Extract and cache the evaluated sources
        const parsedEvaluations = this.parseAndCacheTrustEvaluation(trustEvaluation, uncachedUrls);
        evaluatedSources = parsedEvaluations;
      }
      
      // Combine cached and newly evaluated sources
      const allSources = [...cachedScores, ...evaluatedSources];
      
      // Filter for trusted sources (score > 60) and sort by score
      const trustedSources = allSources
        .filter(source => source.score > 60)
        .sort((a, b) => b.score - a.score);
      
      if (trustedSources.length === 0) {
        return {
          post,
          refusal: 'NOT ENOUGH EVIDENCE TO WRITE A GOOD COMMUNITY NOTE - No sources scored above 60',
          images_summary: imagesSummary
        };
      }

      const mostTrustedUrl = trustedSources[0].url;

      // Phase 4: Fetch content from most trusted source
      let sourceContent = await this.fetchWebContent(mostTrustedUrl);
      
      if (!sourceContent && trustedSources.length > 1) {
        // Try backup trusted source
        sourceContent = await this.fetchWebContent(trustedSources[1].url);
      }

      if (!sourceContent) {
        return {
          post,
          refusal: 'NOT ENOUGH EVIDENCE TO WRITE A GOOD COMMUNITY NOTE - Could not fetch source content',
          images_summary: imagesSummary
        };
      }

      // Phase 5: Verify context in source
      const contextVerification = await this.openRouter.findContextInSource(
        sourceContent,
        mostTrustedUrl,
        missingContext
      );

      let foundValidSource = contextVerification.trim().toUpperCase() === 'YES';
      let validSourceUrl = mostTrustedUrl;

      if (!foundValidSource) {
        // Try other trusted sources (only those scoring > 60)
        for (const source of trustedSources.slice(1, 3)) {
          const altContent = await this.fetchWebContent(source.url);
          if (altContent) {
            const altVerification = await this.openRouter.findContextInSource(
              altContent,
              source.url,
              missingContext
            );
            if (altVerification.trim().toUpperCase() === 'YES') {
              foundValidSource = true;
              validSourceUrl = source.url;
              break;
            }
          }
        }
      }

      // If no source contains the relevant context, refuse to create a note
      if (!foundValidSource) {
        return {
          post,
          refusal: 'NOT ENOUGH EVIDENCE TO WRITE A GOOD COMMUNITY NOTE - Sources do not contain relevant context',
          images_summary: imagesSummary
        };
      }

      // Create the note
      const primaryContext = this.extractPrimaryContext(missingContext);
      const noteText = this.formatNote(primaryContext, validSourceUrl);

      return {
        post,
        note: {
          post_id: post.id,
          note_text: noteText,
          misleading_tags: [MisleadingTag.MissingImportantContext],
          trustworthy_sources: true
        },
        images_summary: imagesSummary
      };

    } catch (error) {
      return {
        post,
        error: `Error in enhanced research pipeline: ${error instanceof Error ? error.message : String(error)}`,
        images_summary: imagesSummary
      };
    }
  }

  private extractUrlsFromClaims(missingContext: string): string[] {
    const urls: string[] = [];
    const lines = missingContext.split('\n');
    let inSourcesSection = false;
    
    for (const line of lines) {
      if (line.trim().toLowerCase() === 'sources:') {
        inSourcesSection = true;
        continue;
      }
      
      if (inSourcesSection && line.trim().startsWith('-')) {
        const urlMatch = line.match(/https?:\/\/[^\s<>"{}|\\^`\[\]]+[^\s<>"{}|\\^`\[\].,;:!?\'")\]]/);
        if (urlMatch) {
          urls.push(urlMatch[0]);
        }
      } else if (inSourcesSection && line.trim() && !line.trim().startsWith('-')) {
        // End of sources section
        inSourcesSection = false;
      }
    }
    
    return [...new Set(urls)]; // Remove duplicates
  }

  private parseAndCacheTrustEvaluation(trustEvaluation: string, urls: string[]): { url: string; score: number; reason: string }[] {
    const evaluatedSources: { url: string; score: number; reason: string }[] = [];
    const lines = trustEvaluation.split('\n');
    
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      
      // Look for lines containing URLs
      for (const url of urls) {
        if (line.includes(url)) {
          // Extract score - look for patterns like "score: 85" or "Trust score: 85" or just "85"
          const scoreMatch = line.match(/(?:score|Score):\s*(\d+)|(?:^|\s)(\d+)(?:\s|$)/);
          if (scoreMatch) {
            const score = parseInt(scoreMatch[1] || scoreMatch[2]);
            
            // Extract reason - usually on the same line or the next line
            let reason = '';
            const reasonMatch = line.match(/(?:reason|Reason):\s*(.+)/);
            if (reasonMatch) {
              reason = reasonMatch[1].trim();
            } else if (i + 1 < lines.length) {
              // Check next line for reason
              reason = lines[i + 1].trim();
            }
            
            // Cache the score
            this.trustCache.setTrustScore(url, score, reason);
            
            evaluatedSources.push({ url, score, reason });
          }
        }
      }
    }
    
    return evaluatedSources;
  }


  private async fetchWebContent(url: string): Promise<string | null> {
    try {
      const response = await axios.get(url, {
        timeout: this.config.fetch_timeout || 10000,  // Default to 10 seconds
        headers: {
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
          'Accept-Encoding': 'gzip, deflate'  // Avoid brotli which Bun doesn't support
        }
      });
      
      // Parse HTML with jsdom
      const dom = new JSDOM(response.data, { url });
      
      // Use Readability to extract article content
      const reader = new Readability(dom.window.document);
      const article = reader.parse();
      
      if (article && article.textContent) {
        // Return clean text content, limited to 30k chars to leave room for prompts
        const cleanText = article.textContent
          .replace(/\s+/g, ' ')  // Normalize whitespace
          .trim()
          .substring(0, 30000);
        
        console.log(`✓ Extracted ${cleanText.length} chars of clean text from ${url}`);
        return cleanText;
      } else {
        console.log(`⚠️  Could not extract article content from ${url}`);
        // Fallback to basic text extraction from body
        const bodyText = dom.window.document.body?.textContent || '';
        return bodyText.replace(/\s+/g, ' ').trim().substring(0, 30000);
      }
    } catch (error) {
      console.error(`Error fetching ${url}:`, error);
      return null;
    }
  }

  private extractPrimaryContext(missingContext: string): string {
    const lines = missingContext.split('\n').filter(line => line.trim());
    let primaryContext = lines[0] || '';
    
    // Remove numbering or bullet points
    primaryContext = primaryContext.replace(/^[\d\-\*\.]+\s*/, '').trim();
    
    return primaryContext;
  }

  private formatNote(context: string, sourceUrl: string): string {
    // Just return the context and source URL without any prefix
    let note = `${context} ${sourceUrl}`;
    
    // Ensure under 280 characters
    if (note.length > 280) {
      const excess = note.length - 280 + 3; // +3 for "..."
      const shortenedContext = context.substring(0, context.length - excess) + '...';
      note = `${shortenedContext} ${sourceUrl}`;
    }
    
    return note;
  }
}